### DISCUSSION TITLE
Pass instance parameters in Setup()

### DISCUSSION THREAD

**Discussion by FVSoares** [lathoub/Arduino-AppleMIDI-Library#139] [General]
I saw that begin can only send midi port number.
Is there a way to send the Session Name and UPD port number inside Setup()?
I would like to let the user choose the session name.
It looks like it was possible in older releases.

Regards

**Comment by lathoub** [comment]
replace the default construction macro `APPLEMIDI_CREATE_DEFAULTSESSION_INSTANCE`, with the parameterised `APPLEMIDI_CREATE_INSTANCE` macro (depending on your ETH hardware) :

https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/a8aceca5e73cd05f80f479391b87e82c22469f9d/src/AppleMIDI.h#L359-L369

The default port is set here

https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/877b39f7593e4375bb6ccb69fb58d3ad5ae34af7/src/AppleMIDI_Defs.h#L13

Or see [this](https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/ae9dabaa12ba7c636d7f89ee3a94cf01e60ec961/examples/AVR_NonDefaultSession/AVR_NonDefaultSession.ino) example

**Reply by FVSoares** [reply]
I think I already did it. I just program for fun so bear with me, I have limited knowledge.
The problem that I have, my apologies I did not describe it well, is that I would like to set the port and session name in the setup() part of Arduino code.
As the instance is created before Setup() I don't know how to change when I can get user interaction.

I believe if I create the instance in Setup() it will be destroyed when the program enters loop().

Does it makes sense?
To be honest my knowledge is limited to look at the source by myself. 
I tried to find a begin call to understand if it is possible to change instance parameters before calling begin, but I did not find how to do it.
I would like the port and session name not to be hard coded inside my code and use a config file to set it up.

Thanks a lot.
Fabio

**Reply by FVSoares** [reply]
Hi again, I found the setName()...
It worked, but I couldn't find anything to set the port.
I will take another look.

**Reply by lathoub** [reply]
Setting the port number in the setup section was not envisioned, as the AppleMIDI protocol implementation is a plugin to the FortySevenEffects MIDI library - with limited wiggle space in what can be called.

Also, I wanted to make it as easy as possible to switch between protocols.

Possible solution: you could add a setPort method

``` cpp
    void setPort(const uint16_t port)
    {
        this->port = port;
    }
```

and call it **_before_** you call `MIDI.begin()`

I have not tested the above, let me know if it works :-)

**Reply by FVSoares** [reply]
port is protected in the library and I wouldn't mess with the library before I learn a bit more.
I think this library is very good and very useful. Change port number is not that important.
Thanks for your support.

**Reply by lathoub** [reply]
> port is protected in the library and I wouldn't mess with the library before I learn a bit more.

You can clone the library in your repo, make the above modif and see how it goes.

When you are ready, pls mark this discussion as "answered"

**Reply by lathoub** [reply]
I added a public method `setPort`, to be called before `MIDI.begin()`

**Reply by lathoub** [reply]
@FVSoares have you been able to test the added `setPort` function?

**Comment by lathoub** [comment]
@FVSoares did this work for you?

**Reply by lathoub** [reply]
@FVSoares gentle nudge, I'm preparing for a new release and want to close items that can be closed

**Comment by FVSoares** [comment]
My apologies I didn`t follow up on this. It can be closed and I tried to
but I don`t know why the button to close it is not available when I log in.

On Fri, Aug 12, 2022 at 3:26 AM lathoub ***@***.***> wrote:

> @FVSoares <https://github.com/FVSoares> gentle nudge, I'm preparing for a
> new release and want to close items that can be closed
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lathoub/Arduino-AppleMIDI-Library/discussions/139#discussioncomment-3381184>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWKGJHZBP75LO4MCHT7EUA3VYXVCXANCNFSM5HHBVRIA>
> .
> You are receiving this because you were mentioned.Message ID:
> <lathoub/Arduino-AppleMIDI-Library/repo-discussions/139/comments/3381184@
> github.com>
>


-- 
Fabio Vieira Soares
Eng. de Sistemas
